Create S3SplitFileInput for reading data out of S3.

mreid-moz opened the pull request:
See [Bug 1126471](https://bugzilla.mozilla.org/show_bug.cgi?id=1126471).
This also adds the "heka-s3list" and "heka-s3cat" commands which can be used together or separately to look at the data in S3.

mreid-moz committed:
Add common code to list/filter an S3 bucket.

mreid-moz committed:
WIP S3 input

mreid-moz committed:
Still WIP. Don't get a single record properly.

If there is only one record in the target file on S3, it doesn't
get parsed out correctly.

mreid-moz committed:
Format code

mreid-moz committed:
Add more logging

mreid-moz committed:
Add common code for reading records from S3 files.

mreid-moz committed:
Progress towards a working S3Input

mreid-moz committed:
Add some s3splitfile tests.

mreid-moz committed:
Add a utility for listing filtered keys in S3.

Given a "schema" specifying the allowable values for each level
of the S3 directory-like structure, output a list of all the
matching keys on stdout.

mreid-moz committed:
Add a utility for getting data from files in S3.

Given a list of filenames as arguments or on stdin, fetch the
records from the specified files. You can use the output of
`heka-s3list` as the input to `heka-s3cat`.

mreid-moz committed:
Update patches to apply cleanly.

mreid-moz committed:
Add a new patch to build heka-s3{cat|list}

mreid-moz committed:
Build more cmds, add arg to select build branch.

If you pass in an argument, it will be used that git branch for
the s3splitfile build.

mreid-moz committed:
Update the s3splitfile import path

mreid-moz committed:
Update code to work with new "splitter" feature.

mreid-moz committed:
Remove commented-out gunk

mreid-moz committed:
Use the new "splitter" mechanism to read from S3.

mreid-moz committed:
Merge branch 'master' into whd_changes

Conflicts:
	bin/build_pipeline_heka.sh

rafrombrc commented on heka/plugins/s3splitfile/s3splitfile_input.go:
```diff
@@ -0,0 +1,203 @@
+
+	deliverer := runner.NewDeliverer(s3Key)
+	defer deliverer.Done()
+	sr := runner.NewSplitterRunner(s3Key)
```
You can create these in `fetcher` and pass them in to `readS3File` to reuse them. Saves us from spinning goroutines up and down each time.

mreid-moz replied:
I just tried doing this, but I think my introduction of the "reachedEOF" field in Heka PR 1305 has changed the reusability of SplitterRunners - they now remember that they're at EOF and therefore don't give you any data after the end of the first stream they see.

mreid-moz replied:
Here's a patch that made the SplitterRunners reusable - just call "Reset()" after each stream. https://gist.github.com/mreid-moz/9f9baa110050789b440a

mreid-moz replied:
With [Heka PR 1317](https://github.com/mozilla-services/heka/pull/1317), we can now safely reuse SplitterRunners on different streams. I've updated the code to do just that - now we create a single SplitterRunner / Deliverer pair per `fetcher` instead of having a pair for each S3 file.

mreid-moz committed:
Run gofmt.

trink commented on heka/cmd/heka-s3cat/main.go:
```diff
@@ -0,0 +1,149 @@
+		} else {
+			if len(record) > 0 {
+				processed += 1
+				headerLen := int(record[1]) + message.HEADER_FRAMING_SIZE
```
The fact that you have to decompose the framing yourself is something we should probably address in the Heka library

mreid-moz replied:
Within the pipeline context you don't need to do it yourself - it's just that these cmds don't have decoders.

trink commented on heka/cmd/heka-s3cat/main.go:
```diff
@@ -0,0 +1,149 @@
+# You can obtain one at http://mozilla.org/MPL/2.0/.
+#
+# The Initial Developer of the Original Code is the Mozilla Foundation.
+# Portions created by the Initial Developer are Copyright (C) 2014
```
We should really not have copyright dates or contributors listed in individual source files (they just end up wrong like above). @rafrombrc

mreid-moz replied:
I'll remove them all, since they're all over the place.  I'll leave the MPL blurb in though, right?

trink commented on heka/cmd/heka-s3list/main.go:
```diff
@@ -0,0 +1,98 @@
+
+	// Initialize the S3 bucket
+	auth := aws.Auth{AccessKey: *flagAWSKey, SecretKey: *flagAWSSecretKey}
+	s := s3.New(auth, aws.Regions[*flagAWSRegion])
```
Why the different handling of region (cat vs list)

mreid-moz replied:
The `cat` handling will give you a sensible error message if you get the parameter wrong. I will update this one to match.

trink commented on heka/patches/0003-Add-more-cmds.patch:
```diff
@@ -0,0 +1,46 @@
+ DEPENDS hekad)
+-- 
+1.9.4 (Apple Git-50.2)
+
```
I'll see what I cat do about getting rid of a bunch of these patch files.

mreid-moz replied:
Thanks! That would be great.

trink commented on heka/plugins/s3splitfile/all_specs_test.go:
```diff
@@ -0,0 +1,30 @@
+#
+# Contributor(s):
+#   Rob Miller (rmiller@mozilla.com)
+#   Mike Trinkala (trink@mozilla.com)
```
Errr, copyright date and contributors

mreid-moz committed:
Print a sensible message on invalid aws region.

trink commented on heka/plugins/s3splitfile/s3splitfile_common.go:
```diff
@@ -214,16 +221,224 @@ func LoadSchema(schemaFileName string) (schema Schema, err error) {
+
+		if err != nil {
+			if err == io.EOF {
+				if len(record) != 0 {
```
If the record was not empty why would it be overwritten with any remaining bytes?

mreid-moz replied:
I wanted to make sure that the Printf call below happened if something weird was going on with EOF and GetRemainingData().  If record _is_ empty, then we don't care what's in GetRemainingData(), since it's an incomplete message.  If it is not empty, we should use it, and _then_ check GetRemainingData (and not overwrite record at this point).  Will fix.

trink commented on heka/plugins/s3splitfile/s3splitfile_input.go:
```diff
@@ -0,0 +1,203 @@
+	pipeline.RegisterPlugin("S3SplitFileInput", func() interface{} {
+		return new(S3SplitFileInput)
+	})
+}
```
It would be nice to have a ReportMsg interface to get some visibility into its operation.

mreid-moz replied:
Filed [bug 1130525](https://bugzilla.mozilla.org/show_bug.cgi?id=1130525) for this.

mreid-moz committed:
Remove contributor info from individual files.

Clean up the contributor information and fix the MPL License
header where needed.

mreid-moz committed:
Do not overwrite a valid record with junk.

Only check if there is remaining data, but do not overwrite the
last record with this (garbage) data.

mreid-moz committed:
Use one deliverer+splitterrunner per fetcher.

This saves spinning up/down a bunch of coroutines per S3 file.

mreid-moz committed:
Check for remaining data in each S3 stream.
